### PR TITLE
Implement SecretBox

### LINKED ISSUE #2: Implement SecretBox
Mostly splitting the work in this PR https://github.com/crossbario/autobahn-java/pull/558 to a proper library

### PR DESCRIPTION
**Pull request by muzzammilshahid**
closed #2

### CODE REVIEW

File: src/main/java/io/xconn/cryptobox/SecretBox.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        }
+
+        byte[] plaintext = new byte[len];
+        xsalsa20.processBytes(ciphertext, poly1305.getMacSize(), plaintext.length, plaintext, 0);
+        return plaintext;
+    }
+
+    private void checkLength(byte[] data, int size) {
+        if (data == null || data.length != size)
+            throw new IllegalArgumentException("Invalid size: " + data.length);
```
Comment (Mahad-10) [inline src/main/java/io/xconn/cryptobox/SecretBox.java]:
```suggestion
            throw new IllegalArgumentException("Invalid size: " + data.length + ". Length should be " + Util.SECRET_KEY_LEN);
```

**Review by Mahad-10 — APPROVED** [review]
LGTM with one suggestion

File: src/main/java/io/xconn/cryptobox/SecretBox.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+import org.bouncycastle.crypto.macs.Poly1305;
+import org.bouncycastle.crypto.params.KeyParameter;
+import org.bouncycastle.crypto.params.ParametersWithIV;
+
+import java.security.MessageDigest;
+import java.util.Arrays;
+
+public class SecretBox {
+
+    static int Poly1305MacSize = 16;
```
Comment (om26er) [inline src/main/java/io/xconn/cryptobox/SecretBox.java]:
```suggestion
    static int MAC_SIZE = 16;
```

File: src/main/java/io/xconn/cryptobox/SecretBox.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        checkLength(key, Util.SECRET_KEY_LEN);
+
+        byte[] nonce = Util.generateRandomBytesArray(Util.NONCE_SIZE);
+        box(output, nonce, message, key);
+    }
+
+    static void box(byte[] output, byte[] nonce, byte[] plaintext, byte[] key) {
+        checkLength(nonce, Util.NONCE_SIZE);
+
+        XSalsa20Engine xsalsa20 = new XSalsa20Engine();
```
Comment (om26er) [inline src/main/java/io/xconn/cryptobox/SecretBox.java]:
```suggestion
        XSalsa20Engine cipher = new XSalsa20Engine();
```

File: src/main/java/io/xconn/cryptobox/SecretBox.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        Poly1305 poly1305 = new Poly1305();
```
Comment (om26er) [inline src/main/java/io/xconn/cryptobox/SecretBox.java]:
```suggestion
        Poly1305 mac = new Poly1305();
```

File: src/main/java/io/xconn/cryptobox/SecretBox.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+        xsalsa20.init(true, new ParametersWithIV(new KeyParameter(key), nonce));
+        byte[] subKey = new byte[Util.SECRET_KEY_LEN];
+        xsalsa20.processBytes(subKey, 0, Util.SECRET_KEY_LEN, subKey, 0);
+        byte[] cipherWithoutNonce = new byte[plaintext.length + poly1305.getMacSize()];
+        xsalsa20.processBytes(plaintext, 0, plaintext.length, cipherWithoutNonce, poly1305.getMacSize());
+
+        // hash ciphertext and prepend mac to ciphertext
```
Comment (om26er) [inline src/main/java/io/xconn/cryptobox/SecretBox.java]:
```suggestion
        // hash the ciphertext
```

File: src/test/java/io/xconn/cryptobox/UtilTest.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+    @Test
+    public void testGenerateRandomBytesArray() {
+        int size = 32;
+        byte[] randomBytes = Util.generateRandomBytesArray(size);
+
+        assertNotNull(randomBytes);
+        assertEquals(size, randomBytes.length);
+    }
+
```
Comment (om26er) [inline src/test/java/io/xconn/cryptobox/UtilTest.java]:
```suggestion
```

**Review by om26er — CHANGES_REQUESTED** [review]
first batch of review

File: src/test/java/io/xconn/cryptobox/SecretBoxTest.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
+package io.xconn.cryptobox;
+
+import static org.junit.jupiter.api.Assertions.assertArrayEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import static io.xconn.cryptobox.SecretBox.MAC_SIZE;
```
Comment (om26er) [inline src/test/java/io/xconn/cryptobox/SecretBoxTest.java:10]: we should order the imports
1. standard library
2. third-party
3. project code

File: src/main/java/io/xconn/cryptobox/SecretBox.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        byte[] message = Arrays.copyOfRange(ciphertext, Util.NONCE_SIZE,
+                ciphertext.length);
+
+        boxOpen(output, nonce, message, key);
+    }
+
+    static void boxOpen(byte[] output, byte[] nonce, byte[] ciphertext, byte[] key) {
+        checkLength(nonce, Util.NONCE_SIZE);
+
+        XSalsa20Engine xsalsa20 = new XSalsa20Engine();
```
Comment (om26er) [inline src/main/java/io/xconn/cryptobox/SecretBox.java]:
```suggestion
        XSalsa20Engine cipher = new XSalsa20Engine();
```

File: src/main/java/io/xconn/cryptobox/SecretBox.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        Poly1305 poly1305 = new Poly1305();
```
Comment (om26er) [inline src/main/java/io/xconn/cryptobox/SecretBox.java]:
```suggestion
        Poly1305 mac = new Poly1305();
```

**Review by om26er — CHANGES_REQUESTED** [review]

File: src/main/java/io/xconn/cryptobox/SecretBox.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+import org.bouncycastle.crypto.params.ParametersWithIV;
+
+import java.security.MessageDigest;
+import java.util.Arrays;
+
+public class SecretBox {
+
+    static int MAC_SIZE = 16;
+
+    public static byte[] box(byte[] message, byte[] key) {
```
Comment (om26er) [inline src/main/java/io/xconn/cryptobox/SecretBox.java]:
```suggestion
    public static byte[] box(byte[] message, byte[] privateKey) {
```

Comment (om26er) [reply src/main/java/io/xconn/cryptobox/SecretBox.java]: here and in all other places

**Review by om26er — CHANGES_REQUESTED** [review]

**Review by om26er — APPROVED** [review]

**Pull request merged**